File: workers/lib/schemas.py
"""Pydantic mirror of docs/contracts.md.

Single source of truth for inter-stage wire formats. Owned by M2 (Architect).

Schema changes follow the process in docs/contracts.md. Don't bypass.
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Annotated, Literal, Protocol

from pydantic import BaseModel, ConfigDict, Field, field_validator

# ─── Common types ─────────────────────────────────────────────────────────────

Color = Annotated[str, Field(pattern=r"^#[0-9A-Fa-f]{6}$", description="Hex color")]
SceneId = Annotated[str, Field(min_length=1, max_length=64)]


class Position(BaseModel):
    model_config = ConfigDict(extra="forbid")
    x: float
    y: float
    z: float = 0.0


# ─── Stage 0 — IngestedDocument ───────────────────────────────────────────────


class Subject(str, Enum):
    MATH = "math"
    CS = "cs"
    PHYSICS = "physics"
    CHEMISTRY = "chemistry"
    BIOLOGY = "biology"
    OTHER = "other"


class Figure(BaseModel):
    model_config = ConfigDict(extra="forbid")
    caption: str
    page: int = Field(ge=1)
    storage_path: str | None = None


class Section(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: str
    level: int = Field(ge=1, le=6)
    title: str
    prose_md: str = ""
    equations: list[str] = []
    figures: list[Figure] = []


class DocumentMetadata(BaseModel):
    model_config = ConfigDict(extra="forbid")
    page_count: int = Field(ge=1, le=50, description="Hard cap 50 in v1")
    language: str = "en"
    detected_subject: Subject | None = None
    source_filename: str
    ingested_at: datetime


class IngestedDocument(BaseModel):
    model_config = ConfigDict(extra="forbid")
    schema_version: int = 1
    title: str
    sections: list[Section]
    metadata: DocumentMetadata


# ─── Stage 1 — CurriculumPlan ─────────────────────────────────────────────────


class Concept(BaseModel):
    model_config = ConfigDict(extra="forbid")
    concept_id: str
    name: str
    prerequisites: list[str] = []


class SceneOutline(BaseModel):
    model_config = ConfigDict(extra="forbid")
    scene_id: SceneId
    concept_id: str
    hook: str
    beat_summary: str
    estimated_sec: float = Field(ge=0)


class CurriculumPlan(BaseModel):
    model_config = ConfigDict(extra="forbid")
    schema_version: int = 1
    target_duration_sec: int = Field(ge=60, le=1800, default=600)
    learning_objectives: list[str]
    concept_dag: list[Concept]
    scene_outline: list[SceneOutline]

    @field_validator("scene_outline")
    @classmethod
    def unique_scene_ids(cls, v: list[SceneOutline]) -> list[SceneOutline]:
        ids = [s.scene_id for s in v]
        if len(ids) != len(set(ids)):
            raise ValueError("scene_outline contains duplicate scene_id")
        return v


# ─── Stage 2 — Script ─────────────────────────────────────────────────────────


class VisualCue(BaseModel):
    model_config = ConfigDict(extra="forbid")
    at_word_index: int = Field(ge=0)
    description: str


class ScriptScene(BaseModel):
    model_config = ConfigDict(extra="forbid")
    scene_id: SceneId
    narration_md: str
    visual_cues: list[VisualCue] = []
    estimated_duration_sec: float = Field(ge=0)

    @field_validator("visual_cues")
    @classmethod
    def cues_strictly_increasing(cls, v: list[VisualCue]) -> list[VisualCue]:
        indices = [c.at_word_index for c in v]
        if indices != sorted(set(indices)):
            raise ValueError("visual_cues must be strictly increasing by at_word_index")
        return v


class Script(BaseModel):
    model_config = ConfigDict(extra="forbid")
    schema_version: int = 1
    scenes: list[ScriptScene]


# ─── Stage 3 — SceneSpec ──────────────────────────────────────────────────────
#
# Elements are a discriminated union by `type`. Codegen depends on exhaustive
# handling of all variants — when you add a new one, also update:
#   - modal/stages/codegen.py element handlers
#   - modal/prompts/scene_spec_v1.md examples
#   - modal/prompts/codegen_v1.md examples


class _ElementBase(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: str
    type: str


class MathTexElement(_ElementBase):
    type: Literal["MathTex"] = "MathTex"
    latex: str
    position: Position = Position(x=0, y=0)
    scale: float = 1.0


class TextElement(_ElementBase):
    type: Literal["Text"] = "Text"
    text: str
    position: Position = Position(x=0, y=0)
    scale: float = 1.0
    font: str | None = None


class AxesElement(_ElementBase):
    type: Literal["Axes"] = "Axes"
    x_range: tuple[float, float, float]
    y_range: tuple[float, float, float]
    position: Position = Position(x=0, y=0)
    scale: float = 1.0


class VectorElement(_ElementBase):
    type: Literal["Vector"] = "Vector"
    components: list[float]
    color: Color = "#FFFFFF"
    anchor_id: str | None = None


class GraphElement(_ElementBase):
    type: Literal["Graph"] = "Graph"
    function: str
    x_range: tuple[float, float]
    axes_id: str


class CircleElement(_ElementBase):
    type: Literal["Circle"] = "Circle"
    radius: float = Field(gt=0)
    position: Position = Position(x=0, y=0)
    color: Color = "#FFFFFF"
    fill_opacity: float = Field(ge=0, le=1, default=0.0)


class GroupElement(_ElementBase):
    type: Literal["Group"] = "Group"
    member_ids: list[str]


class ImageElement(_ElementBase):
    type: Literal["Image"] = "Image"
    storage_path: str
    position: Position = Position(x=0, y=0)
    scale: float = 1.0


Element = Annotated[
    MathTexElement
    | TextElement
    | AxesElement
    | VectorElement
    | GraphElement
    | CircleElement
    | GroupElement
    | ImageElement,
    Field(discriminator="type"),
]


class ActionKind(str, Enum):
    CREATE = "Create"
    FADE_IN = "FadeIn"
    FADE_OUT = "FadeOut"
    TRANSFORM = "Transform"
    INDICATE = "Indicate"
    WRITE = "Write"
    WAIT = "Wait"


class Action(BaseModel):
    model_config = ConfigDict(extra="forbid")
    at_t: float = Field(ge=0)
    duration_sec: float = Field(ge=0)
    action: ActionKind
    target_id: str | None = None
    params: dict = {}

    @field_validator("target_id")
    @classmethod
    def wait_has_no_target(cls, v: str | None, info) -> str | None:
        # Note: at this validator we don't have access to `action` until model_validator;
        # full cross-field check lives in SceneSpec.validate_actions below.
        return v


class SceneSpec(BaseModel):
    model_config = ConfigDict(extra="forbid")
    schema_version: int = 1
    scene_id: SceneId
    duration_sec: float = Field(ge=0)
    background: Color = "#000000"
    elements: list[Element] = []
    timeline: list[Action] = []

    @field_validator("elements")
    @classmethod
    def unique_element_ids(cls, v: list[Element]) -> list[Element]:
        ids = [e.id for e in v]
        if len(ids) != len(set(ids)):
            raise ValueError("elements contains duplicate id")
        return v

    @field_validator("timeline")
    @classmethod
    def actions_well_formed(cls, v: list[Action]) -> list[Action]:
        for a in v:
            if a.action == ActionKind.WAIT and a.target_id is not None:
                raise ValueError(f"Wait action must have target_id=null (got {a.target_id})")
            if a.action != ActionKind.WAIT and a.target_id is None:
                raise ValueError(f"{a.action.value} action requires target_id")
        return v


# ─── Stage 4 — GeneratedScene ─────────────────────────────────────────────────


class GeneratedScene(BaseModel):
    model_config = ConfigDict(extra="forbid")
    schema_version: int = 1
    scene_id: SceneId
    python_source: str
    imports: list[str] = []
    llm_model: str
    llm_attempt: int = Field(ge=1)
    generated_at: datetime


# ─── Stage 5 — RenderedScene ──────────────────────────────────────────────────


class RenderedScene(BaseModel):
    model_config = ConfigDict(extra="forbid")
    schema_version: int = 1
    scene_id: SceneId
    mp4_storage_path: str
    duration_sec: float = Field(ge=0)
    resolution: tuple[int, int] = (1920, 1080)
    is_fallback: bool = False
    render_attempts: int = Field(ge=1, default=1)


# ─── Stage 6 — Audio interface ────────────────────────────────────────────────


class PacingHints(BaseModel):
    model_config = ConfigDict(extra="forbid")
    estimated_duration_sec: float = Field(ge=0)
    pause_at_word_indices: list[int] = []
    emphasis_word_indices: list[int] = []


class NarrationInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    text: str
    voice_id: str | None = None
    pacing: PacingHints


class WordTiming(BaseModel):
    model_config = ConfigDict(extra="forbid")
    word: str
    start_sec: float = Field(ge=0)
    end_sec: float = Field(ge=0)


class AudioOutput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    schema_version: int = 1
    scene_id: SceneId
    audio_storage_path: str
    duration_sec: float = Field(ge=0)
    word_timings: list[WordTiming] = []
    provider: Literal["null", "elevenlabs", "openai"] = "null"


class TTSProvider(Protocol):
    """All TTS implementations satisfy this protocol.

    The pipeline always calls TTSProvider.synthesize() — no conditionals.
    Swap NullTTSProvider for ElevenLabsTTSProvider via config; no other code changes.
    """

    async def synthesize(self, n: NarrationInput, scene_id: SceneId) -> AudioOutput: ...


# ─── Stage 7 — FinalVideo ─────────────────────────────────────────────────────


class FinalVideo(BaseModel):
    model_config = ConfigDict(extra="forbid")
    schema_version: int = 1
    job_id: str
    mp4_storage_path: str
    duration_sec: float
    scene_count: int = Field(ge=1)
    fallback_scene_count: int = Field(ge=0)
    resolution: tuple[int, int] = (1920, 1080)
    has_audio: bool = False
    caption_srt_path: str | None = None


# ─── Job lifecycle (control-plane state, not strictly inter-stage) ────────────


class JobStatus(str, Enum):
    PENDING = "pending"
    INGESTING = "ingesting"
    SCRIPTING = "scripting"
    AWAITING_APPROVAL = "awaiting_approval"
    RENDERING = "rendering"
    COMPOSING = "composing"
    DONE = "done"
    FAILED = "failed"
    CANCELLED = "cancelled"


class JobEventKind(str, Enum):
    STARTED = "started"
    PROGRESS = "progress"
    COMPLETED = "completed"
    ERROR = "error"
    RETRY = "retry"


__all__ = [
    # Stage 0
    "IngestedDocument",
    "Section",
    "Figure",
    "DocumentMetadata",
    "Subject",
    # Stage 1
    "CurriculumPlan",
    "Concept",
    "SceneOutline",
    # Stage 2
    "Script",
    "ScriptScene",
    "VisualCue",
    # Stage 3
    "SceneSpec",
    "Element",
    "MathTexElement",
    "TextElement",
    "AxesElement",
    "VectorElement",
    "GraphElement",
    "CircleElement",
    "GroupElement",
    "ImageElement",
    "Action",
    "ActionKind",
    "Position",
    # Stage 4
    "GeneratedScene",
    # Stage 5
    "RenderedScene",
    # Stage 6
    "TTSProvider",
    "NarrationInput",
    "AudioOutput",
    "PacingHints",
    "WordTiming",
    # Stage 7
    "FinalVideo",
    # Common
    "Color",
    "SceneId",
    "JobStatus",
    "JobEventKind",
]

File: packages/shared/schemas.ts
// Auto-mirrored from modal/lib/schemas.py per docs/contracts.md.
// In production, this is regenerated via `pnpm run codegen:schemas`
// from the Pydantic models. Hand-edited in v1 until codegen pipeline wires up.

// ─── Common ───────────────────────────────────────────────────────────────────

export type Color = string; // pattern: ^#[0-9A-Fa-f]{6}$
export type SceneId = string;

export interface Position {
  x: number;
  y: number;
  z?: number;
}

// ─── Stage 0 — IngestedDocument ──────────────────────────────────────────────

export type Subject = "math" | "cs" | "physics" | "chemistry" | "biology" | "other";

export interface Figure {
  caption: string;
  page: number;
  storage_path: string | null;
}

export interface Section {
  id: string;
  level: number;
  title: string;
  prose_md: string;
  equations: string[];
  figures: Figure[];
}

export interface DocumentMetadata {
  page_count: number;
  language: string;
  detected_subject: Subject | null;
  source_filename: string;
  ingested_at: string; // ISO datetime
}

export interface IngestedDocument {
  schema_version: number;
  title: string;
  sections: Section[];
  metadata: DocumentMetadata;
}

// ─── Stage 1 — CurriculumPlan ────────────────────────────────────────────────

export interface Concept {
  concept_id: string;
  name: string;
  prerequisites: string[];
}

export interface SceneOutline {
  scene_id: SceneId;
  concept_id: string;
  hook: string;
  beat_summary: string;
  estimated_sec: number;
}

export interface CurriculumPlan {
  schema_version: number;
  target_duration_sec: number;
  learning_objectives: string[];
  concept_dag: Concept[];
  scene_outline: SceneOutline[];
}

// ─── Stage 2 — Script ────────────────────────────────────────────────────────

export interface VisualCue {
  at_word_index: number;
  description: string;
}

export interface ScriptScene {
  scene_id: SceneId;
  narration_md: string;
  visual_cues: VisualCue[];
  estimated_duration_sec: number;
}

export interface Script {
  schema_version: number;
  scenes: ScriptScene[];
}

// ─── Stage 3 — SceneSpec ─────────────────────────────────────────────────────

interface ElementBase {
  id: string;
  type: string;
}

export interface MathTexElement extends ElementBase {
  type: "MathTex";
  latex: string;
  position: Position;
  scale: number;
}
export interface TextElement extends ElementBase {
  type: "Text";
  text: string;
  position: Position;
  scale: number;
  font: string | null;
}
export interface AxesElement extends ElementBase {
  type: "Axes";
  x_range: [number, number, number];
  y_range: [number, number, number];
  position: Position;
  scale: number;
}
export interface VectorElement extends ElementBase {
  type: "Vector";
  components: number[];
  color: Color;
  anchor_id: string | null;
}
export interface GraphElement extends ElementBase {
  type: "Graph";
  function: string;
  x_range: [number, number];
  axes_id: string;
}
export interface CircleElement extends ElementBase {
  type: "Circle";
  radius: number;
  position: Position;
  color: Color;
  fill_opacity: number;
}
export interface GroupElement extends ElementBase {
  type: "Group";
  member_ids: string[];
}
export interface ImageElement extends ElementBase {
  type: "Image";
  storage_path: string;
  position: Position;
  scale: number;
}

export type Element =
  | MathTexElement
  | TextElement
  | AxesElement
  | VectorElement
  | GraphElement
  | CircleElement
  | GroupElement
  | ImageElement;

export type ActionKind =
  | "Create"
  | "FadeIn"
  | "FadeOut"
  | "Transform"
  | "Indicate"
  | "Write"
  | "Wait";

export interface Action {
  at_t: number;
  duration_sec: number;
  action: ActionKind;
  target_id: string | null;
  params: Record<string, unknown>;
}

export interface SceneSpec {
  schema_version: number;
  scene_id: SceneId;
  duration_sec: number;
  background: Color;
  elements: Element[];
  timeline: Action[];
}

// ─── Stage 5 — RenderedScene ─────────────────────────────────────────────────

export interface RenderedScene {
  schema_version: number;
  scene_id: SceneId;
  mp4_storage_path: string;
  duration_sec: number;
  resolution: [number, number];
  is_fallback: boolean;
  render_attempts: number;
}

// ─── Stage 7 — FinalVideo ────────────────────────────────────────────────────

export interface FinalVideo {
  schema_version: number;
  job_id: string;
  mp4_storage_path: string;
  duration_sec: number;
  scene_count: number;
  fallback_scene_count: number;
  resolution: [number, number];
  has_audio: boolean;
  caption_srt_path: string | null;
}

// ─── Job lifecycle ───────────────────────────────────────────────────────────

export type JobStatus =
  | "pending"
  | "ingesting"
  | "scripting"
  | "awaiting_approval"
  | "rendering"
  | "composing"
  | "done"
  | "failed"
  | "cancelled";

export type JobEventKind = "started" | "progress" | "completed" | "error" | "retry";

export interface Job {
  id: string;
  user_id: string;
  status: JobStatus;
  pdf_storage_path: string;
  target_duration_sec: number;
  approved: boolean;
  final_video_path: string | null;
  error_message: string | null;
  cost_estimate_usd: number | null;
  created_at: string;
  updated_at: string;
}

export interface JobEvent {
  id: number;
  job_id: string;
  stage: string;
  kind: JobEventKind;
  payload: Record<string, unknown>;
  created_at: string;
}
